clippy_lints/src/methods/useless_nonzero_new_unchecked.rs

```diff
@@ -0,0 +1,61 @@
+use clippy_utils::diagnostics::{span_lint_and_sugg, span_lint_and_then};
+use clippy_utils::is_in_const_context;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Block, BlockCheckMode, Expr, ExprKind, Node, QPath, UnsafeSource};
+use rustc_hir_analysis::lower_ty;
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::USELESS_NONZERO_NEW_UNCHECKED;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>, func: &Expr<'tcx>, args: &[Expr<'_>], msrv: &Msrv) {
+    if msrv.meets(msrvs::CONST_UNWRAP)
+        && let ExprKind::Path(QPath::TypeRelative(ty, segment)) = func.kind
+        && segment.ident.name == sym::new_unchecked
+        && let [init_arg] = args
+        && let ExprKind::Lit(_) = init_arg.kind
+        && is_in_const_context(cx)
```

Comment:
Shouldn't this be restricted to `const` and `static` items specifically (i.e. `is_inside_always_const_context()`), rather than any const context, as only those are actually guaranteed to evaluate at compile time? `const fn`s may still execute at runtime so there could technically be a performance difference at least w/o optimization. The way I understood the linked issue is that it should lint when there would really be no difference between the two at runtime. (Otherwise, if you're relying on optimizations either way, then is the const const check really needed?)

Reply:
I thought that `const` functions called with `const` arguments (such as literals, which is the only case linted here) were guaranteed to evaluate at compile time. Is that not the case? (I may well be wrong)

Reply:
Indeed. I will tighten the check.

Reply:
Done.

Reply:
Since we are not inside a compile time evaluation, I removed the check for literals, as it applies to any argument passed to `unchecked_new()`.

Reply:
Thanks for catching this, I went through the PR looking for this issue and was pleased to see it handled correctly already!